### PR TITLE
Apex log integration(Initial commit)

### PR DESCRIPTION
**Pull request by ChimanJain**

### CODE REVIEW

File: apexlog/apexlog.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+	"github.com/airbrake/gobrake/v5"
+
+	"github.com/apex/log"
+)
+
+// severity identifies the sort of log: info, warning etc. It also implements
+// the flag.Value interface. The -stderrthreshold flag is of type severity and
+// should be modified only through the flag.Value interface. The values match
+// the corresponding constants in C++.
```
Comment (kyrylo) [inline apexlog/apexlog.go:16]: Please remove the C++ mention. This looks irrelevant.

File: apexlog/apexlog.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/airbrake/gobrake/v5"
+
+	"github.com/apex/log"
+)
+
+// severity identifies the sort of log: info, warning etc. It also implements
+// the flag.Value interface. The -stderrthreshold flag is of type severity and
```
Comment (kyrylo) [inline apexlog/apexlog.go:14]: * How do we use this flag with this integration?
* Why does it need to implement the `flag.Value` interface?

File: apexlog/apexlog.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		}
+	}
+	return 0, false
+}
+
+var gobrakeSeverity = errorLog
+
+// SetGobrakeSeverity sets minimum log severity that will be sent to Airbrake.
+//
+// Valid names are "debug" "info", "warning", "error", and "fatal".  If the name is not
```
Comment (kyrylo) [inline apexlog/apexlog.go:51]: Why do we allow a situation when the name is not recognized? Instead of accepting `severity string` in `setGobrakeSeverity` we can accept type `severity`, therefore this:

```go
func setGobrakeSeverity(severity string) {
```

...becomes:

```go
func setGobrakeSeverity(s severity) {
```

Then we won't have to validate during runtime.

File: apexlog/apexlog.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+// recognized, "error" severity is used.
+func setGobrakeSeverity(severity string) {
```
Comment (kyrylo) [inline apexlog/apexlog.go:53]: This name is misleading because it doesn't actually set Gobrake's severity (the library, or the notifier). It just sets state in our handler.

File: apexlog/apexlog.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	if Gobrake == nil {
+		return
+	}
+	if s < gobrakeSeverity {
+		return
+	}
+
+	msg := fmt.Sprint(args...)
+
+	var theErr error
```
Comment (kyrylo) [inline apexlog/apexlog.go:107]: I have never seen this name being used in Go. Can we use the standard name `err` instead?

File: examples/apexlog/main.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	})
+
+	// Close the airbrake notifier
+	defer airbrake.Close()
+
+	// Create the airbrake handler for "github.com/apex/log"
+	// Insert the airbrake notifier instance and set the severity level.
+	// The acceptable severity levels are "debug" "info", "warning", "error", and "fatal".
+	// If the severity level is not recognized, "error" severity is used.
+	airbrakeHandler := apexlog.New(airbrake, "error")
```
Comment (kyrylo) [inline examples/apexlog/main.go:27]: We shouldn't accepts strings, it should be `severity` so that you don't validate input during runtime.

```suggestion
	airbrakeHandler := apexlog.New(airbrake, apexlog.ErrorLevel)
```

File: examples/apexlog/main.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+func main() {
+	// Initialize the airbrake notifier
+	airbrake := gobrake.NewNotifierWithOptions(&gobrake.NotifierOptions{
+		ProjectId:   363389,                             // use it as an environment variable
+		ProjectKey:  "baa755018d5e35e07897ee0087fcce9c", // use it as an environment variable
+		Environment: "production",                       // use it as an environment variable
+	})
+
+	// Close the airbrake notifier
```
Comment (kyrylo) [inline examples/apexlog/main.go:20]: This comment doesn't look useful because it doesn't tell anything new.

File: examples/apexlog/main.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	"errors"
+	"time"
+
+	"github.com/airbrake/gobrake/v5"
+	"github.com/airbrake/gobrake/v5/apexlog"
+	"github.com/apex/log"
+)
+
+func main() {
+	// Initialize the airbrake notifier
```
Comment (kyrylo) [inline examples/apexlog/main.go:13]: This comment doesn't look useful because it doesn't tell anything new.

File: apexlog/apexlog.go [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+	notice.Context["severity"] = severityName[s]
+
+	if values != nil {
+		route, _ := values.Value("route").(string)
+		if route != "" {
+			notice.Context["route"] = route
+		}
+	}
+
+	Gobrake.SendNoticeAsync(notice)
```
Comment (kyrylo) [inline apexlog/apexlog.go:142]: We should stick to the standard `Notify` API. Here's example from the docs:

```go
notice := gobrake.NewNotice("operation failed", nil, 0)
notice.Context["severity"] = "critical"
airbrake.Notify(notice, nil)
```

File: apexlog/apexlog.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	var notice *gobrake.Notice
+	if theErr != nil {
+		notice = Gobrake.Notice(theErr, req, depth)
+		notice.Errors[0].Message = msg
+	} else {
+		notice = Gobrake.Notice(msg, req, depth)
+	}
+	notice.Context["severity"] = severityName[s]
```
Comment (kyrylo) [inline apexlog/apexlog.go:133]: If `s` implemented `String()`, we could call:
```suggestion
	notice.Context["severity"] = s.String()
```

File: apexlog/apexlog.go [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+			values = v
+		}
+	}
+
+	if values != nil && req == nil {
+		req, _ = values.Value("request").(*http.Request)
+	}
+
+	var notice *gobrake.Notice
+	if theErr != nil {
```
Comment (kyrylo) [inline apexlog/apexlog.go:127]: Since you use `if else` ... you can flip the branches so that we don't negate without a good reason. It's just easier to read `if this... then that` vs `if not this... then that`.

File: apexlog/apexlog.go [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+		return
+	}
+	if s < gobrakeSeverity {
+		return
+	}
+
+	msg := fmt.Sprint(args...)
+
+	var theErr error
+	var req *http.Request
```
Comment (kyrylo) [inline apexlog/apexlog.go:108]: Why do we mess with HTTP requests in an integration with a logging library?

File: examples/apexlog/main.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+	}()
+
+	go func() {
+		for range time.Tick(2 * time.Second) {
+			err := errors.New("boom")
+			ctx.WithError(err).Error("upload failed")
+		}
+	}()
+
+	select {}
```
Comment (kyrylo) [inline examples/apexlog/main.go:66]: This example looks rather complex IMO. What we could do is to:

1. Set severity to `warning`
2. Send a debug log and explain it won't be delivered
3. Send a fatal log and explain that it will be delivered
4. Set severity to `debug`
5. Send a debug log and explain that it will be delivered

No need for goroutines.

File: apexlog/apexlog.go [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+// New Apex Logs handler for airbrake.
+//
+// Valid severity levels are "debug" "info", "warning", "error", and "fatal".  If the name is not
+// recognized, "error" severity is used.
+func New(notifier *gobrake.Notifier, severity string) *Handler {
+	setGobrakeSeverity(severity)
+	var v Handler
+	Gobrake = notifier
```
Comment (kyrylo) [inline apexlog/apexlog.go:76]: What will happen if two projects use this `apexlog` integration? Probably only one integration will win and the other one will stop sending logs. Can we make sure we support this scenario?

I am actually not sure why we store `notifier` on package-level. We store it on the handler itself on the next line, so why not to use it from there?

**Review by kyrylo — CHANGES_REQUESTED** [review]

File: apexlog/apexlog.go (hunk 4 above)
Comment (ChimanJain) [reply apexlog/apexlog.go:53]: changed the name.

File: apexlog/apexlog.go (hunk 5 above)
Comment (ChimanJain) [reply apexlog/apexlog.go:107]: done

File: examples/apexlog/main.go (hunk 7 above)
Comment (ChimanJain) [reply examples/apexlog/main.go:20]: It is to generalise the example, so the people don't get confused if it requires different configuration.

File: examples/apexlog/main.go (hunk 8 above)
Comment (ChimanJain) [reply examples/apexlog/main.go:13]: Removed it.

Comment (kyrylo) [reply examples/apexlog/main.go:13]: This is still present.

File: examples/apexlog/main.go (hunk 6 above)
Comment (kyrylo) [reply examples/apexlog/main.go:27]: We still pass a string instead of `severity`.

File: apexlog/apexlog.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+	errorLog: "error",
+	fatalLog: "fatal",
+}
+
+// String implementation for severity.
+func (s severity) String() string {
+	return severityName[s]
+}
+
+func parseSeverity(s string) (severity, bool) {
```
Comment (kyrylo) [inline apexlog/apexlog.go:39]: We don't need this function if `severity` is of type `severity`, not `string`.

File: apexlog/apexlog.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+// HandleLog method is used for sending notices to airbrake.
+func (h *Handler) HandleLog(e *log.Entry) error {
+	h.notifyAirbrake(0, severity(e.Level), e.Message)
+
+	return nil
+}
+
+func (h *Handler) notifyAirbrake(depth int, s severity, arg interface{}) {
+	if h.Gobrake == nil {
```
Comment (kyrylo) [inline apexlog/apexlog.go:85]: I am not sure this condition will ever be `true`. We have to call `New` prior to calling `notifyAirbrake`, which sets `h.Gobrake`.

File: apexlog/apexlog.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+func New(notifier *gobrake.Notifier, severity string) *Handler {
+	setHandlerSeverity(severity)
+	var v Handler
+	v.Gobrake = notifier
+	return &v
+}
+
+// HandleLog method is used for sending notices to airbrake.
+func (h *Handler) HandleLog(e *log.Entry) error {
+	h.notifyAirbrake(0, severity(e.Level), e.Message)
```
Comment (kyrylo) [inline apexlog/apexlog.go:79]: We don't need to have `0` (depth) as a function parameter because it never changes. Let's just remove it and hardcode it to 0 on line 93.

File: examples/apexlog/main.go [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+	// Set the airbrake handler in the log
+	log.SetHandler(airbrakeHandler)
+
+	ctx := log.WithFields(log.Fields{
+		"file": "something.png",
+		"type": "image/png",
+		"user": "tobi",
+	})
+
+	for range time.Tick(time.Millisecond * 500) {
```
Comment (kyrylo) [inline examples/apexlog/main.go:40]: Why do we need a timer?

File: examples/apexlog/main.go [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+		ctx.Info("upload")
+		ctx.Info("upload complete")
+		ctx.Warn("upload retry")
+		ctx.WithError(errors.New("unauthorized")).Error("upload failed")
+		ctx.Errorf("failed to upload %s", "img.png")
+	}
+}
```
Comment (kyrylo) [inline examples/apexlog/main.go:47]: It would be nice to print a message that encourages the user to check their project on Airbrake after the logs were sent.

**Review by kyrylo — CHANGES_REQUESTED** [review]

File: apexlog/apexlog.go (hunk 16 above)
Comment (ChimanJain) [reply apexlog/apexlog.go:85]: removed it.

File: examples/apexlog/main.go (hunk 18 above)
Comment (ChimanJain) [reply examples/apexlog/main.go:40]: This example is use used for the apex/log package(the one they used in their package), so we can show that the middleware will work without making any modifications to apex/log and it will be easier for developers to understand who are using apex/log already.

File: examples/apexlog/main.go (hunk 19 above)
Comment (ChimanJain) [reply examples/apexlog/main.go:47]: Please explain it briefly. Not sure what to write.

File: apexlog/apexlog.go (hunk 17 above)
Comment (ChimanJain) [reply apexlog/apexlog.go:79]: It's just a variable, so that we can build on it in future if we want to or if we want to add some another variable and need some reference.

File: apexlog/apexlog.go (hunk 15 above)
Comment (ChimanJain) [reply apexlog/apexlog.go:39]: I didn't wanted to expose the severity variable in the project, it was to make sure no one can modify it or may lead to memory leakage issue. It needs to be inside the package, if you want we can define constants that can be used outside the package.

Comment (kyrylo) [reply apexlog/apexlog.go:39]: On line 18 you can export constants outside the package. Then the users will be able to pass severity like this:

```
apexlog.New(notifier, apexlog.DebugLog)
```

I don't see where we can leak memory. It's basically how the `log` package works. In this PR we have:

```
log.SetLevel(log.DebugLevel)
```

That's the same as `apexlog.New(notifier, apexlog.DebugLog)`

File: apexlog/apexlog.go (hunk 17 above)
Comment (kyrylo) [reply apexlog/apexlog.go:79]: Could you provide an example of a possible scenario in the future where we want to pass anything other than 0 here?

File: examples/apexlog/main.go (hunk 19 above)
Comment (kyrylo) [reply examples/apexlog/main.go:47]:
```go
fmt.Println("Go check your Airbrake dashboard at https://airbrake.io/projects/123 to see these log messages")
```

**Comment by kyrylo** [comment]
Is this ready for another review?

**Comment by ChimanJain** [comment]
Yes

**Comment by kyrylo** [comment]
Ok, thanks. Please write "PTAL" once you are done with the code. This will help me understand the state of the PR 🙂

File: examples/apexlog/main.go (hunk 18 above)
Comment (kyrylo) [reply examples/apexlog/main.go:43]: I gave this a run locally and I sent 160 errors to Airbrake within ~15 seconds. I don't think it's a good idea to run this on a loop because our users can leave this running for some time and then they will run out of their error quota or will have to pay more than needed.

I think we need to print these messages without a loop and make the program exit by itself so that we won't waste error quota.

File: examples/apexlog/main.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+	// Set the airbrake handler in the log
+	log.SetHandler(airbrakeHandler)
+
+	ctx := log.WithFields(log.Fields{
+		"file": "something.png",
+		"type": "image/png",
+		"user": "tobi",
+	})
+
+	fmt.Println("Check your Airbrake dashboard at https://airbrake.io/projects/:ProjectId to see these log messages")
```
Comment (kyrylo) [inline examples/apexlog/main.go:42]: * We should print the real project id being used so that users can actually click on the link
* It would be nice to explain that some of the log messages won't be delivered to Airbrake due to severity settings. As of now this message implies all of the messages will be delivered, which isn't actually the case

File: apexlog/apexlog.go [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+	ErrorLog
+	FatalLog
+)
+
+var severityName = []string{
+	DebugLog: "debug",
+	InfoLog:  "info",
+	WarnLog:  "warn",
+	ErrorLog: "error",
+	FatalLog: "fatal",
```
Comment (kyrylo) [inline apexlog/apexlog.go:30]: When I send `ctx.Fatal("oops")`, it exits the program but never reports anything to Airbrake.

Comment (ChimanJain) [reply apexlog/apexlog.go:30]: The apex/log package exits the application internally denying us to log it in airbrake

Comment (kyrylo) [reply apexlog/apexlog.go:30]: If we cannot support this, we shouldn't be exposing this severity in our integration because it doesn't do anything useful. In doing so, we should probably document this behavior. If I set up `FataLog` severity, nothing at all would be sent to Airbrake, which is misleading (I would think that only fatal logs would be sent).

Comment (ChimanJain) [reply apexlog/apexlog.go:30]: Ok, I will remove that.

**Comment by ChimanJain** [comment]
PTAL

**Comment by kyrylo** [comment]
Did you see these comments?

* https://github.com/airbrake/gobrake/pull/220#discussion_r758226011
* https://github.com/airbrake/gobrake/pull/220#discussion_r758226742

They don't appear to be resolved.

**Review by kyrylo — APPROVED** [review]

**Comment by ChimanJain** [comment]
PTAL

**Review by kyrylo — APPROVED** [review]
